### PR TITLE
Introduce missing MQMD fields to the `ibmmq:Message`

### PR DESCRIPTION
**Pull request by ayeshLK**
Fixes: [#6964](https://github.com/ballerina-platform/ballerina-library/issues/6964)
Related to: [#775](https://github.com/wso2-enterprise/internal-support-ballerina/issues/775)

### Using `userId` and `accountingToken`

```ballerina
// Open the Queue/Topic with `ibmmq:MQOO_SET_IDENTITY_CONTEXT` option enabled
Queue queue = check queueManager.accessQueue("DEV.QUEUE.1", .... | ibmmq:MQOO_SET_IDENTITY_CONTEXT);

string accountingToken = "<accounting-token>";
string userId = "<userId>";
check queue->put({
    accountingToken: accountingToken.toBytes(),
    userId,
    payload: "Hello world!".toBytes()
}, options = ibmmq:MQPMO_SET_IDENTITY_CONTEXT);
```

### Using `encoding` and `characterSet`

```ballerina
json messageBody = {
    user: "Doe, John",
     id: 1234,
     description: "Regular application developer"
};
byte[] payload = messageBody.toJsonString().toBytes();
check queue->put({
    characterSet: ibmmq:MQCCSI_UTF8,
    encoding: ibmmq:MQENC_INTEGER_NORMAL,
    payload
});
```

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec
- [ ] Checked native-image compatibility

### CODE REVIEW

File: ballerina/constants.bal [hunk 1]
```diff
@@ -104,6 +104,64 @@ public const int MQGMO_CONVERT = 16384;
...
+public const int MQCCSI_ASCII_ISO = 819;
+
+# Coded Character Set Identifiers - As Published.
+public const int MQCCSI_AS_PUBLISHED = -4;
+
+# The CodedCharSetId of the data in the String field is defined by the CodedCharSetId field in the header structure 
+# that precedes the MQCFH structure, or by the CodedCharSetId field in the MQMD if the MQCFH is at the start of the message.
+public const int MQCCSI_DEFAULT = 0;
+
+# The American EBCDIC codeset
```
Comment (MohamedSabthar) [inline ballerina/constants.bal:144]:
```suggestion
# The American EBCDIC codeset.
```

File: ballerina/constants.bal [hunk 2]
```diff
@@ -104,6 +104,64 @@ public const int MQGMO_CONVERT = 16384;
...
+# Encoding for reversed integer representation (least significant byte first) or the little-endian format.
+public const int MQENC_INTEGER_REVERSED = 2;
+
+# Packed-decimal integers are represented in the conventional way: Each decimal digit in the printable form of the number is represented in packed decimal by a single hexadecimal digit in the range X'0' through X'9'.
+public const int MQENC_DECIMAL_NORMAL = 16;
+
+# Packed-decimal integers are represented in the same way as `ENC_DECIMAL_NORMAL`, but with the bytes arranged in reverse order.
+public const int MQENC_DECIMAL_REVERSED = 32;
+
+# Floating-point numbers are represented using the standard IEEE3 floating-point format
```
Comment (MohamedSabthar) [inline ballerina/constants.bal:119]:
```suggestion
# Floating-point numbers are represented using the standard IEEE3 floating-point format.
```

File: ballerina/constants.bal [hunk 3]
```diff
@@ -104,6 +104,64 @@ public const int MQGMO_CONVERT = 16384;
...
+# Floating-point numbers are represented in the same way as `ENC_FLOAT_IEEE_NORMAL`, but with the bytes arranged in reverse order.
+public const int MQENC_FLOAT_IEEE_REVERSED = 512;
+
+# Floating-point numbers are represented using the standard zSeries (System/390) floating-point format.
+public const int MQENC_FLOAT_S390 = 768;
+
+# Coded Character Set Identifiers - Appl.
+public const int  MQCCSI_APPL = -3;
+
+# ASCII codeset
```
Comment (MohamedSabthar) [inline ballerina/constants.bal:131]:
```suggestion
# ASCII codeset.
```

File: ballerina/constants.bal [hunk 4]
```diff
@@ -104,6 +104,64 @@ public const int MQGMO_CONVERT = 16384;
...
+public const int MQCCSI_ASCII = 850;
+
+# ISO standard ASCII codeset
```
Comment (MohamedSabthar) [inline ballerina/constants.bal:134]:
```suggestion
# ISO standard ASCII codeset.
```

Comment (MohamedSabthar) [reply ballerina/constants.bal:134]: check other places

File: ballerina/tests/queue_producer_consumer_tests.bal [hunk 5]
```diff
@@ -518,6 +518,110 @@ function produceAndConsumerMessageWithMultipleHeaderTypesWithJsonPayloadTest() r
...
+@test:Config {
+    groups: ["ibmmqQueue", "messageIdentification"]
+}
+function produceMessagesWithIdentification() returns error? {
+    QueueManager queueManager = check new (
+        name = "QM1", host = "localhost", channel = "DEV.ADMIN.SVRCONN", 
+        userID = "admin", password = "password");
+    Queue queue = check queueManager.accessQueue("DEV.QUEUE.1", MQOO_OUTPUT | MQOO_INPUT_AS_Q_DEF | MQOO_SET_IDENTITY_CONTEXT);
+
+    string messageContent = "This is a sample message with a identification.";
```
Comment (NipunaRanasinghe) [inline ballerina/tests/queue_producer_consumer_tests.bal:530]:
```suggestion
    string messageContent = "This is a sample message with an identification.";
```

File: ballerina/tests/queue_producer_consumer_tests.bal [hunk 6]
```diff
@@ -518,6 +518,110 @@ function produceAndConsumerMessageWithMultipleHeaderTypesWithJsonPayloadTest() r
...
+    check queueManager.disconnect();
+}
+
+function trimTrailingZeros(byte[] bytes) returns byte[] {
+    int i = bytes.length() - 1;
+    while (i >= 0 && bytes[i] == 0) {
+        i -= 1;
+    }
+    return bytes.slice(0, i + 1);
+}
```
Comment (NipunaRanasinghe) [inline ballerina/tests/queue_producer_consumer_tests.bal:558]: shall we move the test utils to the end of the file?

File: ballerina/constants.bal (hunk 4 above)
Comment (ayeshLK) [reply ballerina/constants.bal]: Fixed with [947ddab1c60decba2e6cfd3225dfad3662e1f619](https://github.com/ballerina-platform/module-ballerinax-ibm.ibmmq/pull/56/commits/947ddab1c60decba2e6cfd3225dfad3662e1f619)

File: ballerina/constants.bal (hunk 3 above)
Comment (ayeshLK) [reply ballerina/constants.bal]: Fixed with [947ddab1c60decba2e6cfd3225dfad3662e1f619](https://github.com/ballerina-platform/module-ballerinax-ibm.ibmmq/pull/56/commits/947ddab1c60decba2e6cfd3225dfad3662e1f619)

File: ballerina/constants.bal (hunk 2 above)
Comment (ayeshLK) [reply ballerina/constants.bal]: Fixed with [947ddab1c60decba2e6cfd3225dfad3662e1f619](https://github.com/ballerina-platform/module-ballerinax-ibm.ibmmq/pull/56/commits/947ddab1c60decba2e6cfd3225dfad3662e1f619)

**Review by MohamedSabthar — APPROVED** [review]

File: ballerina/constants.bal [hunk 7]
```diff
@@ -104,6 +104,64 @@ public const int MQGMO_CONVERT = 16384;
 # Subscribe Option create
 public const int MQSO_CREATE = 2;
 
+# Encoding for normal integer representation (most significant byte first) or the big-endian format.
+public const int MQENC_INTEGER_NORMAL = 1;
```
Comment (NipunaRanasinghe) [inline ballerina/constants.bal:108]: type descriptor is redundant for constants. any reason not to omit that?
```suggestion
public const MQENC_INTEGER_NORMAL = 1;
```

Comment (ayeshLK) [reply ballerina/constants.bal:108]: We have used the same practice for other constants defined in the package. We will follow the same practice for time-being to maintain the consistency.

**Review by NipunaRanasinghe — APPROVED** [review]
LGTM

**Pull request merged**